### PR TITLE
refactor: isolate metrics tests

### PR DESCRIPTION
**Pull request by xcoulon**
Fixes https://issues.redhat.com/browse/CRT-1179

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: test/metrics/metrics_test.go [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+				t.Run("verify metrics", func(t *testing.T) {
+					metricsAssertion.WaitForMetricDelta(UserSignupsMetric, 2)                                            // no change
+					metricsAssertion.WaitForMetricDelta(UserSignupsApprovedMetric, 4)                                    // two more because of reactivated user
+					metricsAssertion.WaitForMetricDelta(UserSignupsDeactivatedMetric, 2)                                 // no change
+					metricsAssertion.WaitForMetricDelta(MasterUserRecordsPerDomainMetric, 1, "domain", "external")       // userSignupMember1
+					metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 1, "cluster_name", memberAwait.ClusterName)  // userSignupMember1 is on member1
+					metricsAssertion.WaitForMetricDelta(MasterUserRecordsPerDomainMetric, 1, "domain", "external")       // userSignupMember2
+					metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 1, "cluster_name", member2Await.ClusterName) // userSignupMember2 is on member2
+				})
+			})
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:132]: this looks like a duplication of https://github.com/codeready-toolchain/toolchain-e2e/blob/900ccf701b2ab7e5bfc51f95fa91e2dac24562b7/test/metrics/metrics_test.go#L428-L461

File: test/metrics/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+			// when deleting user "user-0002"
+			err = hostAwait.Client.Delete(context.TODO(), usersignups["user-0002"])
+
+			// then
+			require.NoError(t, err)
+			// and verify that the values of the `sandbox_users_per_activations` metric
+			metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 0, "activations", "1", "domain", "external") // (same offset as above)
+			metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 0, "activations", "2", "domain", "external") // user-0002 has been deleted but metric remains unchanged
+			metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 0, "activations", "3", "domain", "external") // (unchanged after other usersignup was deleted)
+
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:506]: having these two cases `restart host-operator pod and verify that metrics are still available` & `delete usersignups` as sub-tests of `when users reactivated` is a bit confusing - they don't have anything in common with reactivation - they only reuse the users

File: test/metrics/metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,735 @@
...
+					metricsAssertion.WaitForMetricDelta(UserSignupsApprovedMetric, 1) // user approved
+					metricsAssertion.WaitForMetricDelta(UserSignupsBannedMetric, 0)   // unchanged: banneduser already existed
+					metricsAssertion.WaitForMetricDelta(MasterUserRecordsPerDomainMetric, 1, "domain", "external")
+					metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 1, "cluster_name", memberAwait.ClusterName)  // user provisioned on member1
+					metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", member2Await.ClusterName) // no user on member2
+				})
+			})
+		})
+
+		t.Run("manually created usersignup with preexisting banneduser", func(t *testing.T) {
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:581]: this is also more likely a test of the UserSignup controller logic, and not of the metrics

File: test/metrics/metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,735 @@
...
+				metricsAssertion.WaitForMetricDelta(UserSignupsMetric, 1)
+				metricsAssertion.WaitForMetricDelta(UserSignupsApprovedMetric, 0) // not provisioned because banned before signup
+				metricsAssertion.WaitForMetricDelta(UserSignupsBannedMetric, 1)
+				metricsAssertion.WaitForMetricDelta(MasterUserRecordsPerDomainMetric, 0, "domain", "external")
+				metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", memberAwait.ClusterName)  // no user on member1
+				metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", member2Await.ClusterName) // no user on member2
+			})
+		})
+
+		t.Run("register new user with preexisting ban", func(t *testing.T) {
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:622]: the same also in this case - we don't verify metrics logic here, but reg-service

File: test/metrics/metrics_test.go [hunk 5]
```diff
@@ -0,0 +1,735 @@
...
+				metricsAssertion.WaitForMetricDelta(UserSignupsApprovedMetric, 0) // not provisioned because banned before signup
+				metricsAssertion.WaitForMetricDelta(UserSignupsBannedMetric, 0)   // banneduser resource was created before signup
+				metricsAssertion.WaitForMetricDelta(MasterUserRecordsPerDomainMetric, 0, "domain", "external")
+				metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", memberAwait.ClusterName)  // no user on member1
+				metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", member2Await.ClusterName) // no user on member2
+			})
+		})
+	})
+
+	t.Run("when user disabled", func(t *testing.T) {
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:662]: it's again the same thing - there is even no code that would check the metrics

**Review by MatousJobanek — COMMENTED** [review]
Apart from what I wrote in the code I have a few additional comments:
1. you did not change the `go test` command so the new `metrics_test.go` file is not executed
2. the number of sub-tests in `metrics_test.go` is pretty huge and it's really hard to follow the code - please split them into multiple tests
3. try to add `// given` `// when` `// then` comments
4. I don't see any reason for putting the verification of the metrics into an additional sub-test like `t.Run("verify metrics", func(t *testing.T) {`
5. when you drop unrelated tests (as I commented in the code), please make sure that you verify both internal & external usersignup domains

**Comment by xcoulon** [comment]
> Apart from what I wrote in the code I have a few additional comments:
> 
> 1. you did not change the `go test` command so the new `metrics_test.go` file is not executed
> 2. the number of sub-tests in `metrics_test.go` is pretty huge and it's really hard to follow the code - please split them into multiple tests
> 3. try to add `// given` `// when` `// then` comments
> 4. I don't see any reason for putting the verification of the metrics into an additional sub-test like `t.Run("verify metrics", func(t *testing.T) {`
> 5. when you drop unrelated tests (as I commented in the code), please make sure that you verify both internal & external usersignup domains

my apologies for the poor quality of this code. I did a quick duplication of the metric tests but I should have spent more time polishing the code :/ I'll fix that today

**Comment by xcoulon** [comment]
@MatousJobanek I believe I addressed all your comments with https://github.com/codeready-toolchain/toolchain-e2e/pull/342/commits/ffbfb74b14fc0a3e011067ee674d6e7bfa46e904
Could you please take another look at this PR?

File: test/metrics/metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+		_, err = hostAwait.WaitForRouteToBeAvailable(hostAwait.Namespace, "host-operator-metrics-service", "/metrics")
+		require.NoError(t, err, "failed while setting up or waiting for the route to the 'host-operator-metrics-service' service to be available")
+		// also verify that the metric values "survived" the restart
+		metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 0, "activations", "1", "domain", "external") // user-0001 was 1 time (unchanged after pod restarted)
+		metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 0, "activations", "2", "domain", "external") // user-0002 was 2 times (unchanged after pod restarted)
+		metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 0, "activations", "3", "domain", "external") // user-0003 was 3 times (unchanged after pod restarted)
+	})
+}
+
+func TestMetricsWhenUsersDeleted(t *testing.T) {
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:81]: could we rename it for example to:
```suggestion
func TestUsersPerActivationsMetricsStaysTheSameWhenUsersDeleted(t *testing.T) {
```

File: test/metrics/metrics_test.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+			require.NoError(t, err)
+			// reactivate the user
+			CreateAndApproveSignup(t, hostAwait, username, memberAwait.ClusterName, WithIdentityID(usersignups[username].Spec.Userid))
+		}
+	}
+
+	// then verify the value of the `sandbox_users_per_activations` metric
+	metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 1, "activations", "1", "domain", "external") // 1 activation
+	metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 1, "activations", "2", "domain", "external") // 1 activation
+	metricsAssertion.WaitForMetricDelta(UsersPerActivationsAndDomainMetric, 1, "activations", "3", "domain", "external") // 1 activation
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:60]: you could also verify the deactivation metric here. And probably also the others like `UserSignupsMetric` `UserSignupsApprovedMetric` ...

File: test/metrics/metrics_test.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+		// verify the metrics
+		metricsAssertion.WaitForMetricDelta(UserSignupsMetric, 1)
+		metricsAssertion.WaitForMetricDelta(UserSignupsApprovedMetric, 0) // not provisioned because banned before signup
+		metricsAssertion.WaitForMetricDelta(UserSignupsBannedMetric, 1)
+		metricsAssertion.WaitForMetricDelta(MasterUserRecordsPerDomainMetric, 0, "domain", "external")
+		metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", memberAwait.ClusterName)  // no user on member1
+		metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", member2Await.ClusterName) // no user on member2
+	})
+
+	t.Run("register new user with preexisting ban", func(t *testing.T) {
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:205]: I think that you can drop these two cases - it's a duplication of what is alraedy in the `user_management_test.go` and there is nothing related to metrics that wouldn't be already tested before, right?

File: test/metrics/metrics_test.go [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+		metricsAssertion.WaitForMetricDelta(UserSignupsApprovedMetric, 0) // not provisioned because banned before signup
+		metricsAssertion.WaitForMetricDelta(UserSignupsBannedMetric, 0)   // banneduser resource was created before signup
+		metricsAssertion.WaitForMetricDelta(MasterUserRecordsPerDomainMetric, 0, "domain", "external")
+		metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", memberAwait.ClusterName)  // no user on member1
+		metricsAssertion.WaitForMetricDelta(UserAccountsMetric, 0, "cluster_name", member2Await.ClusterName) // no user on member2
+	})
+
+}
+
+func TestMetricsWhenUserDisabled(t *testing.T) {
```
Comment (MatousJobanek) [inline test/metrics/metrics_test.go:241]: let's rename this as well to know what is the purpose of the test
```suggestion
func TestWhenUserDisabledThenNoImpactOnMetrics(t *testing.T) {
```

**Review by MatousJobanek — COMMENTED** [review]
I'm afraid that you didn't cover some cases:
* I didn't see any test that would verify the deactivation metric
* in all test cases you use only the external domain, but not internal one
* none of the tests verify creation in member-2 cluster

File: make/test.mk [hunk 10]
```diff
@@ -68,6 +68,9 @@ e2e-run:
 	oc get toolchaincluster -n $(MEMBER_NS)
 	MEMBER_NS=${MEMBER_NS} MEMBER_NS_2=${MEMBER_NS_2} HOST_NS=${HOST_NS} REGISTRATION_SERVICE_NS=${REGISTRATION_SERVICE_NS} go test ./test/e2e -v -timeout=90m -failfast || \
 	($(MAKE) print-logs HOST_NS=${HOST_NS} MEMBER_NS=${MEMBER_NS} MEMBER_NS_2=${MEMBER_NS_2} REGISTRATION_SERVICE_NS=${REGISTRATION_SERVICE_NS} && exit 1)
+	echo "running the metrics tests"
+	MEMBER_NS=${MEMBER_NS} MEMBER_NS_2=${MEMBER_NS_2} HOST_NS=${HOST_NS} REGISTRATION_SERVICE_NS=${REGISTRATION_SERVICE_NS} go test ./test/metrics -v -timeout=90m -failfast || \
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply make/test.mk:72]: it doesn't make much difference, does it?

File: test/metrics/metrics_test.go (hunk 6 above)
Comment (xcoulon) [reply test/metrics/metrics_test.go:81]: I'm not a fan of such long names TBH

File: test/metrics/metrics_test.go (hunk 8 above)
Comment (xcoulon) [reply test/metrics/metrics_test.go:205]: I don't get it: are we supposed to verify the metrics or not? the goal of this PR is to move all metrics assertions out of the other test cases, so when do you want to check this?

**Comment by xcoulon** [comment]
> I'm afraid that you didn't cover some cases:
> 
> * I didn't see any test that would verify the deactivation metric
> * in all test cases you use only the external domain, but not internal one
> * none of the tests verify creation in member-2 cluster

all the cases above were addressed in 84b2868

File: make/test.mk (hunk 10 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply make/test.mk:72]: much less of code duplication

File: test/metrics/metrics_test.go (hunk 6 above)
Comment (MatousJobanek) [reply test/metrics/metrics_test.go:128]: it's self-explanatory.... 
or you could add a comment to the test.
either a comment or the self-explanatory test name would really help :pray:

File: test/metrics/metrics_test.go (hunk 8 above)
Comment (MatousJobanek) [reply test/metrics/metrics_test.go]: >  so when do you want to check this?

check what? what kind of metrics functionality do you test here and wasn't already tested before in `ban provisioned usersignup`?

File: make/test.mk (hunk 10 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply make/test.mk:72]: ok, fine, let me try that

File: test/metrics/metrics_test.go (hunk 8 above)
Comment (xcoulon) [reply test/metrics/metrics_test.go]: ok, I'll remove this sub-test if you believe it's useless in the scope of metric assertions

Comment (xcoulon) [reply test/metrics/metrics_test.go]: done in aab86da

File: test/metrics/metrics_test.go (hunk 6 above)
Comment (xcoulon) [reply test/metrics/metrics_test.go:128]: done in aab86da

File: make/test.mk (hunk 10 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply make/test.mk:72]: done in aab86da

**Comment by xcoulon** [comment]
@MatousJobanek when you have time, could you take another look at this PR, please?

**Review by MatousJobanek — APPROVED** [review]
Looks good. Thanks a lot for taking extra time and addressing my comments :pray:

File: make/test.mk [hunk 11]
```diff
@@ -66,7 +66,7 @@ test-e2e-registration-local:
 e2e-run:
 	oc get toolchaincluster -n $(HOST_NS)
 	oc get toolchaincluster -n $(MEMBER_NS)
-	MEMBER_NS=${MEMBER_NS} MEMBER_NS_2=${MEMBER_NS_2} HOST_NS=${HOST_NS} REGISTRATION_SERVICE_NS=${REGISTRATION_SERVICE_NS} go test ./test/e2e -v -timeout=90m -failfast || \
+	MEMBER_NS=${MEMBER_NS} MEMBER_NS_2=${MEMBER_NS_2} HOST_NS=${HOST_NS} REGISTRATION_SERVICE_NS=${REGISTRATION_SERVICE_NS} go test ./test/e2e ./test/metrics -v -timeout=90m -failfast || \
```
Comment (MatousJobanek) [inline make/test.mk:69]: don't run them in parallel https://medium.com/@xcoulon/how-to-avoid-parallel-execution-of-tests-in-golang-763d32d88eec :smile: 
```suggestion
	MEMBER_NS=${MEMBER_NS} MEMBER_NS_2=${MEMBER_NS_2} HOST_NS=${HOST_NS} REGISTRATION_SERVICE_NS=${REGISTRATION_SERVICE_NS} go test ./test/e2e ./test/metrics -p 1 -v -timeout=90m -failfast || \
```
